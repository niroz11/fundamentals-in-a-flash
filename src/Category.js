import React, { Component } from 'react';

class Category extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showQuestionOptions: false
    };
  }

  clearStorangeAndSetupQuizWithAllQs = (event) => {
    event.preventDefault();
    localStorage.removeItem(this.props.category);
    this.props.setupQuiz(this.props.category);
  }

  retrieveCorrectAnswerIds = (category) => {
    return JSON.parse(localStorage.getItem(category))
  }

  setupQuizWithoutCorrectlyAnsweredQs = (event) => {
    event.preventDefault();
    let correctlyAnsweredIds = this.retrieveCorrectAnswerIds(this.props.category);
    this.props.setupQuiz(this.props.category, correctlyAnsweredIds)
  }

  render() {
    let { setupQuiz, category, questionsPerCategory } = this.props;
    let style = category.toLowerCase().split(' ').join('-');
    let allCategoryQs = questionsPerCategory[category];
    let numMastered = this.retrieveCorrectAnswerIds(category).length;
    return (
      <div className={"category-contain " + style}>
        <p className="category">{category}</p>
        <p>You have mastered {numMastered} of {allCategoryQs} questions, great work!</p>
        <p>How would you like to quiz your knowledge today?</p>
        <div className="btns-contain">
          <button className="small-btn" onClick={this.clearStorangeAndSetupQuizWithAllQs}>Try all questions</button>
          {
            (numMastered !== allCategoryQs) && <button className="small-btn" onClick={this.setupQuizWithoutCorrectlyAnsweredQs}>Try mastering the other {allCategoryQs - numMastered}</button>
          }  
        </div>
      </div>
    )
  }
}

export default Category;