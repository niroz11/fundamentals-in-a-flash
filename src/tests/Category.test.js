import React from 'react';
import Category from '../Category';
import { shallow } from 'enzyme';

const setupQuizMock = jest.fn();
const index = 0;
const category = 'Prototype Methods';
const questionsPerCategory = {
  "Prototype Methods": 12, 
  "OOP": 8, 
  "Scope": 10 
};


describe('Category', () => {
  let wrapper;

  beforeEach(() => {
    localStorage.setItem('Prototype Methods', "[4, 5, 7]")
    wrapper = shallow(
      <Category
      setupQuiz={setupQuizMock} 
      category={category} 
      questionsPerCategory={questionsPerCategory}
      key={index} 
      />
    );
  });

  it('should match the snapshot with all data passed in', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should retrieve the correct answer ids', () => {
    let correctAnswerIds = wrapper.instance().retrieveCorrectAnswerIds('Prototype Methods');
    expect(correctAnswerIds).toEqual([4, 5, 7]);
  });

  it('should send the correct answer ids and category to app when setupQuizWithoutCorrectlyAnsweredQs is called', () => {
    wrapper.instance().setupQuizWithoutCorrectlyAnsweredQs({ preventDefault: () => { } });
    expect(setupQuizMock).toHaveBeenCalledWith(category, [4, 5, 7]);
  });

  it('should remove the category from localStorage if there and pass the category to app when clearStorangeAndSetupQuizWithAllQs is called', () => {
    expect(localStorage.hasOwnProperty('Prototype Methods')).toEqual(true);
    wrapper.instance().clearStorangeAndSetupQuizWithAllQs({preventDefault: () => {}});
    expect(localStorage.hasOwnProperty('Prototype Methods')).toEqual(false);
    expect(setupQuizMock).toHaveBeenCalledWith(category)
  });

  it('should send the correct ids, category and string of one to app when setupQuizWithOneQuestion has been called', () => {
    localStorage.setItem('Prototype Methods', "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]")
    wrapper.instance().setupQuizWithOneQuestion({ preventDefault: () => { } });
    expect(setupQuizMock).toHaveBeenCalledWith(category, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 'one');    
  });

});