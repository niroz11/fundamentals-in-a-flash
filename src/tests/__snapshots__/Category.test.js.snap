// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category should match the snapshot with all data passed in 1`] = `
<div
  className="category-contain prototype-methods"
>
  <p
    className="category"
  >
    Prototype Methods
  </p>
  <p>
    You have mastered 
    3
     of 
    12
     questions, great work!
  </p>
  <p>
    How would you like to quiz your knowledge today?
  </p>
  <div
    className="btns-contain"
  >
    <button
      className="small-btn"
      onClick={[Function]}
    >
      Try all questions
    </button>
    <button
      className="small-btn"
      onClick={[Function]}
    >
      Try mastering the other 
      9
    </button>
  </div>
</div>
`;
